{
 "cells": [
  {
   "cell_type": "raw",
   "id": "39efcba3-2196-4966-82bc-663fc4f4d3dc",
   "metadata": {
    "tags": []
   },
   "source": [
    "Loading the files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "99a69167-97d1-4861-bb63-bb984886b88b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import pandas as pd\n",
    "input_file=\"/g/data/te53/sj2852/blgrp/analyses/pangenome/analysis-v2/labelled2chm13-blast2anchors/bldgrp.flank.nodup.paf\"\n",
    "column_names = ['geneanchor', 'contig', 'identity', 'length', 'mismatches', 'gaps', 'qstart', 'qend', 'tstart', 'tend', 'evalue', 'bitscore']\n",
    "paf = pd.read_csv(input_file, sep='\\t', header=None, names=column_names)\n",
    "paf = paf[['geneanchor','contig','identity', 'tstart' , 'tend']]\n",
    "fai=\"/g/data/te53/sj2852/blgrp/analyses/pangenome/analysis-v2/labelled2chm13-blast2anchors/bldgrp.genes.combined.fasta.fai\"\n",
    "column_names = ['contig','length','x','y','z']\n",
    "faidf = pd.read_csv(fai, sep='\\t', header=None, names=column_names)\n",
    "faidf = faidf[['contig','length']]"
   ]
  },
  {
   "cell_type": "raw",
   "id": "29d20331-fb92-4e2d-86de-339178b7a067",
   "metadata": {},
   "source": [
    "Merging DBs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9136e44c-3b07-4d11-911b-cae7701cdc7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.merge(paf,faidf,on=['contig'])\n",
    "df[['anchorgene', 'pos']] = df['geneanchor'].str.split(pat='-', n=1, expand=True)\n",
    "df[['sample', 'hap', 'contig', 'contiggene']] = df['contig'].str.split('#', n=3, expand=True)\n",
    "df= df[['anchorgene','pos','sample', 'hap', 'contig', 'identity', 'contiggene','tstart','tend','length']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bf752822-5943-4d85-82f7-ba7f66457b34",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>anchorgene</th>\n",
       "      <th>pos</th>\n",
       "      <th>sample</th>\n",
       "      <th>hap</th>\n",
       "      <th>contig</th>\n",
       "      <th>identity</th>\n",
       "      <th>contiggene</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>GM19467</td>\n",
       "      <td>2</td>\n",
       "      <td>60.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>191956</td>\n",
       "      <td>192156</td>\n",
       "      <td>288477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>GM19471</td>\n",
       "      <td>2</td>\n",
       "      <td>21.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>165882</td>\n",
       "      <td>166082</td>\n",
       "      <td>274006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>GM19473</td>\n",
       "      <td>1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>118198</td>\n",
       "      <td>118398</td>\n",
       "      <td>280985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>GM19473</td>\n",
       "      <td>2</td>\n",
       "      <td>6.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>117969</td>\n",
       "      <td>118169</td>\n",
       "      <td>280754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>HG00097</td>\n",
       "      <td>1</td>\n",
       "      <td>47.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>107008</td>\n",
       "      <td>107208</td>\n",
       "      <td>212619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44346</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>PGXX22539</td>\n",
       "      <td>1</td>\n",
       "      <td>9.0</td>\n",
       "      <td>99.005</td>\n",
       "      <td>XK</td>\n",
       "      <td>30322</td>\n",
       "      <td>30521</td>\n",
       "      <td>105697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44347</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>PGXX22539</td>\n",
       "      <td>2</td>\n",
       "      <td>9.0</td>\n",
       "      <td>99.005</td>\n",
       "      <td>XK</td>\n",
       "      <td>30314</td>\n",
       "      <td>30512</td>\n",
       "      <td>105714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44348</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>PGXX22540</td>\n",
       "      <td>2</td>\n",
       "      <td>27.1</td>\n",
       "      <td>99.005</td>\n",
       "      <td>XK</td>\n",
       "      <td>33788</td>\n",
       "      <td>33986</td>\n",
       "      <td>66780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44349</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>PLPN239131</td>\n",
       "      <td>1</td>\n",
       "      <td>38.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>XK</td>\n",
       "      <td>16390</td>\n",
       "      <td>16590</td>\n",
       "      <td>44399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44350</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>PLPN239131</td>\n",
       "      <td>2</td>\n",
       "      <td>38.1</td>\n",
       "      <td>99.010</td>\n",
       "      <td>XK</td>\n",
       "      <td>16381</td>\n",
       "      <td>16581</td>\n",
       "      <td>44413</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44351 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      anchorgene pos      sample hap contig  identity contiggene  tstart  \\\n",
       "0         A4GALT   3     GM19467   2   60.0   100.000     A4GALT  191956   \n",
       "1         A4GALT   3     GM19471   2   21.0   100.000     A4GALT  165882   \n",
       "2         A4GALT   3     GM19473   1    6.0   100.000     A4GALT  118198   \n",
       "3         A4GALT   3     GM19473   2    6.0   100.000     A4GALT  117969   \n",
       "4         A4GALT   3     HG00097   1   47.0   100.000     A4GALT  107008   \n",
       "...          ...  ..         ...  ..    ...       ...        ...     ...   \n",
       "44346         XK   5   PGXX22539   1    9.0    99.005         XK   30322   \n",
       "44347         XK   5   PGXX22539   2    9.0    99.005         XK   30314   \n",
       "44348         XK   5   PGXX22540   2   27.1    99.005         XK   33788   \n",
       "44349         XK   5  PLPN239131   1   38.1   100.000         XK   16390   \n",
       "44350         XK   5  PLPN239131   2   38.1    99.010         XK   16381   \n",
       "\n",
       "         tend  length  \n",
       "0      192156  288477  \n",
       "1      166082  274006  \n",
       "2      118398  280985  \n",
       "3      118169  280754  \n",
       "4      107208  212619  \n",
       "...       ...     ...  \n",
       "44346   30521  105697  \n",
       "44347   30512  105714  \n",
       "44348   33986   66780  \n",
       "44349   16590   44399  \n",
       "44350   16581   44413  \n",
       "\n",
       "[44351 rows x 10 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "raw",
   "id": "0a3c816e-149e-46a0-ab59-d98e054c18a4",
   "metadata": {},
   "source": [
    "Checking if anchor genes are always the same as contig genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0172b1f7-e19c-40c8-a306-c987df226dfc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>anchorgene</th>\n",
       "      <th>pos</th>\n",
       "      <th>sample</th>\n",
       "      <th>hap</th>\n",
       "      <th>contig</th>\n",
       "      <th>identity</th>\n",
       "      <th>contiggene</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>PBXP050271</td>\n",
       "      <td>2</td>\n",
       "      <td>16.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>REFERENCE</td>\n",
       "      <td>871</td>\n",
       "      <td>1071</td>\n",
       "      <td>18246</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>PBYP331517</td>\n",
       "      <td>2</td>\n",
       "      <td>1.3</td>\n",
       "      <td>100.000</td>\n",
       "      <td>REFERENCE</td>\n",
       "      <td>1767</td>\n",
       "      <td>1967</td>\n",
       "      <td>20554</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>PBYP338524</td>\n",
       "      <td>1</td>\n",
       "      <td>55.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>SLC4A1</td>\n",
       "      <td>45500</td>\n",
       "      <td>45700</td>\n",
       "      <td>61462</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>384</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>PBYP338524</td>\n",
       "      <td>2</td>\n",
       "      <td>55.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>SLC4A1</td>\n",
       "      <td>45474</td>\n",
       "      <td>45674</td>\n",
       "      <td>61436</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>PBYP378545</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>REFERENCE</td>\n",
       "      <td>556</td>\n",
       "      <td>756</td>\n",
       "      <td>15902</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44064</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>HG02675</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>99.502</td>\n",
       "      <td>B3GALNT1</td>\n",
       "      <td>298045</td>\n",
       "      <td>298244</td>\n",
       "      <td>299725</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44109</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>HG03611</td>\n",
       "      <td>1</td>\n",
       "      <td>10.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>GYPA_GYPB</td>\n",
       "      <td>148692</td>\n",
       "      <td>148492</td>\n",
       "      <td>198991</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44110</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>HG03611</td>\n",
       "      <td>2</td>\n",
       "      <td>10.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>GYPA_GYPB</td>\n",
       "      <td>148752</td>\n",
       "      <td>148552</td>\n",
       "      <td>199074</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44144</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>HG03897</td>\n",
       "      <td>1</td>\n",
       "      <td>57.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>SLC14A1</td>\n",
       "      <td>63096</td>\n",
       "      <td>63296</td>\n",
       "      <td>302700</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44145</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>HG03897</td>\n",
       "      <td>2</td>\n",
       "      <td>57.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>SLC14A1</td>\n",
       "      <td>63082</td>\n",
       "      <td>63282</td>\n",
       "      <td>302732</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3072 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      anchorgene pos      sample hap contig  identity contiggene  tstart  \\\n",
       "303       A4GALT   3  PBXP050271   2   16.0   100.000  REFERENCE     871   \n",
       "379       A4GALT   3  PBYP331517   2    1.3   100.000  REFERENCE    1767   \n",
       "383       A4GALT   3  PBYP338524   1   55.0   100.000     SLC4A1   45500   \n",
       "384       A4GALT   3  PBYP338524   2   55.0   100.000     SLC4A1   45474   \n",
       "398       A4GALT   3  PBYP378545   1    1.0   100.000  REFERENCE     556   \n",
       "...          ...  ..         ...  ..    ...       ...        ...     ...   \n",
       "44064         XK   5     HG02675   2    1.0    99.502   B3GALNT1  298045   \n",
       "44109         XK   5     HG03611   1   10.0   100.000  GYPA_GYPB  148692   \n",
       "44110         XK   5     HG03611   2   10.0   100.000  GYPA_GYPB  148752   \n",
       "44144         XK   5     HG03897   1   57.0   100.000    SLC14A1   63096   \n",
       "44145         XK   5     HG03897   2   57.0   100.000    SLC14A1   63082   \n",
       "\n",
       "         tend  length  gene_match  \n",
       "303      1071   18246       False  \n",
       "379      1967   20554       False  \n",
       "383     45700   61462       False  \n",
       "384     45674   61436       False  \n",
       "398       756   15902       False  \n",
       "...       ...     ...         ...  \n",
       "44064  298244  299725       False  \n",
       "44109  148492  198991       False  \n",
       "44110  148552  199074       False  \n",
       "44144   63296  302700       False  \n",
       "44145   63282  302732       False  \n",
       "\n",
       "[3072 rows x 11 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['gene_match'] = df['anchorgene'] == df['contiggene']\n",
    "df[df['gene_match'] == False]"
   ]
  },
  {
   "cell_type": "raw",
   "id": "a5250f5d-a7d0-44e4-9368-16c2ad123308",
   "metadata": {},
   "source": [
    "Checking if the orientation is incorrect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c8e6d58b-b530-4535-a82d-038a85b965f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>anchorgene</th>\n",
       "      <th>pos</th>\n",
       "      <th>sample</th>\n",
       "      <th>hap</th>\n",
       "      <th>contig</th>\n",
       "      <th>identity</th>\n",
       "      <th>contiggene</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_match</th>\n",
       "      <th>strand_match</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>GM19467</td>\n",
       "      <td>2</td>\n",
       "      <td>60.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>191956</td>\n",
       "      <td>192156</td>\n",
       "      <td>288477</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>GM19471</td>\n",
       "      <td>2</td>\n",
       "      <td>21.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>165882</td>\n",
       "      <td>166082</td>\n",
       "      <td>274006</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>GM19473</td>\n",
       "      <td>1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>118198</td>\n",
       "      <td>118398</td>\n",
       "      <td>280985</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>GM19473</td>\n",
       "      <td>2</td>\n",
       "      <td>6.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>117969</td>\n",
       "      <td>118169</td>\n",
       "      <td>280754</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>HG00097</td>\n",
       "      <td>1</td>\n",
       "      <td>47.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>107008</td>\n",
       "      <td>107208</td>\n",
       "      <td>212619</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44346</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>PGXX22539</td>\n",
       "      <td>1</td>\n",
       "      <td>9.0</td>\n",
       "      <td>99.005</td>\n",
       "      <td>XK</td>\n",
       "      <td>30322</td>\n",
       "      <td>30521</td>\n",
       "      <td>105697</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44347</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>PGXX22539</td>\n",
       "      <td>2</td>\n",
       "      <td>9.0</td>\n",
       "      <td>99.005</td>\n",
       "      <td>XK</td>\n",
       "      <td>30314</td>\n",
       "      <td>30512</td>\n",
       "      <td>105714</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44348</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>PGXX22540</td>\n",
       "      <td>2</td>\n",
       "      <td>27.1</td>\n",
       "      <td>99.005</td>\n",
       "      <td>XK</td>\n",
       "      <td>33788</td>\n",
       "      <td>33986</td>\n",
       "      <td>66780</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44349</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>PLPN239131</td>\n",
       "      <td>1</td>\n",
       "      <td>38.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>XK</td>\n",
       "      <td>16390</td>\n",
       "      <td>16590</td>\n",
       "      <td>44399</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44350</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>PLPN239131</td>\n",
       "      <td>2</td>\n",
       "      <td>38.1</td>\n",
       "      <td>99.010</td>\n",
       "      <td>XK</td>\n",
       "      <td>16381</td>\n",
       "      <td>16581</td>\n",
       "      <td>44413</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44351 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      anchorgene pos      sample hap contig  identity contiggene  tstart  \\\n",
       "0         A4GALT   3     GM19467   2   60.0   100.000     A4GALT  191956   \n",
       "1         A4GALT   3     GM19471   2   21.0   100.000     A4GALT  165882   \n",
       "2         A4GALT   3     GM19473   1    6.0   100.000     A4GALT  118198   \n",
       "3         A4GALT   3     GM19473   2    6.0   100.000     A4GALT  117969   \n",
       "4         A4GALT   3     HG00097   1   47.0   100.000     A4GALT  107008   \n",
       "...          ...  ..         ...  ..    ...       ...        ...     ...   \n",
       "44346         XK   5   PGXX22539   1    9.0    99.005         XK   30322   \n",
       "44347         XK   5   PGXX22539   2    9.0    99.005         XK   30314   \n",
       "44348         XK   5   PGXX22540   2   27.1    99.005         XK   33788   \n",
       "44349         XK   5  PLPN239131   1   38.1   100.000         XK   16390   \n",
       "44350         XK   5  PLPN239131   2   38.1    99.010         XK   16381   \n",
       "\n",
       "         tend  length  gene_match  strand_match  \n",
       "0      192156  288477        True          True  \n",
       "1      166082  274006        True          True  \n",
       "2      118398  280985        True          True  \n",
       "3      118169  280754        True          True  \n",
       "4      107208  212619        True          True  \n",
       "...       ...     ...         ...           ...  \n",
       "44346   30521  105697        True          True  \n",
       "44347   30512  105714        True          True  \n",
       "44348   33986   66780        True          True  \n",
       "44349   16590   44399        True          True  \n",
       "44350   16581   44413        True          True  \n",
       "\n",
       "[44351 rows x 12 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['strand_match'] = df['tend'] > df['tstart']\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "85d12529-8a55-4764-ae5d-40fd1f6964cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "duplicate_mappings = df.groupby(['anchorgene', 'pos', 'sample', 'hap']).size().reset_index(name='Count')\n",
    "\n",
    "# Step 2: Filter rows where count is greater than 1 (duplicates)\n",
    "duplicates = duplicate_mappings[duplicate_mappings['Count'] > 1]\n",
    "\n",
    "# Step 3: Merge back with original data to get the identity values for these duplicates\n",
    "duplicate_with_identity = pd.merge(duplicates, df, on=['anchorgene', 'pos', 'sample', 'hap'])\n",
    "\n",
    "# Step 4: Update the 'newheader' and 'oldheader' columns\n",
    "# For non-duplicate tuples, set the newheader and oldheader based on sample, hap, contig, and anchorgene/contiggene\n",
    "# For duplicates, keep the oldheader unchanged.\n",
    "\n",
    "# First, create the newheader and oldheader for non-duplicate entries\n",
    "df['newheader'] = df['sample'] + '#' + df['hap'].astype(str) + '#' + df['contig'].astype(str) + '#' + df['anchorgene']\n",
    "df['oldheader'] = df['sample'] + '#' + df['hap'].astype(str) + '#' + df['contig'].astype(str) + '#' + df['contiggene']\n",
    "\n",
    "# Now, keep oldheader for duplicate entries\n",
    "#df.loc[df.set_index(['anchorgene', 'pos', 'sample', 'hap']).index.isin(duplicates.set_index(['anchorgene', 'pos', 'sample', 'hap']).index), 'newheader'] = df['oldheader']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5e949245-2a59-4eac-ad98-534d31688d39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>anchorgene</th>\n",
       "      <th>pos</th>\n",
       "      <th>sample</th>\n",
       "      <th>hap</th>\n",
       "      <th>contig</th>\n",
       "      <th>identity</th>\n",
       "      <th>contiggene</th>\n",
       "      <th>tstart</th>\n",
       "      <th>tend</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_match</th>\n",
       "      <th>strand_match</th>\n",
       "      <th>newheader</th>\n",
       "      <th>oldheader</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>GM19467</td>\n",
       "      <td>2</td>\n",
       "      <td>60.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>191956</td>\n",
       "      <td>192156</td>\n",
       "      <td>288477</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>GM19467#2#60.0#A4GALT</td>\n",
       "      <td>GM19467#2#60.0#A4GALT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>GM19471</td>\n",
       "      <td>2</td>\n",
       "      <td>21.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>165882</td>\n",
       "      <td>166082</td>\n",
       "      <td>274006</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>GM19471#2#21.0#A4GALT</td>\n",
       "      <td>GM19471#2#21.0#A4GALT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>GM19473</td>\n",
       "      <td>1</td>\n",
       "      <td>6.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>118198</td>\n",
       "      <td>118398</td>\n",
       "      <td>280985</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>GM19473#1#6.0#A4GALT</td>\n",
       "      <td>GM19473#1#6.0#A4GALT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>GM19473</td>\n",
       "      <td>2</td>\n",
       "      <td>6.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>117969</td>\n",
       "      <td>118169</td>\n",
       "      <td>280754</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>GM19473#2#6.0#A4GALT</td>\n",
       "      <td>GM19473#2#6.0#A4GALT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>3</td>\n",
       "      <td>HG00097</td>\n",
       "      <td>1</td>\n",
       "      <td>47.0</td>\n",
       "      <td>100.000</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>107008</td>\n",
       "      <td>107208</td>\n",
       "      <td>212619</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>HG00097#1#47.0#A4GALT</td>\n",
       "      <td>HG00097#1#47.0#A4GALT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44346</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>PGXX22539</td>\n",
       "      <td>1</td>\n",
       "      <td>9.0</td>\n",
       "      <td>99.005</td>\n",
       "      <td>XK</td>\n",
       "      <td>30322</td>\n",
       "      <td>30521</td>\n",
       "      <td>105697</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>PGXX22539#1#9.0#XK</td>\n",
       "      <td>PGXX22539#1#9.0#XK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44347</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>PGXX22539</td>\n",
       "      <td>2</td>\n",
       "      <td>9.0</td>\n",
       "      <td>99.005</td>\n",
       "      <td>XK</td>\n",
       "      <td>30314</td>\n",
       "      <td>30512</td>\n",
       "      <td>105714</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>PGXX22539#2#9.0#XK</td>\n",
       "      <td>PGXX22539#2#9.0#XK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44348</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>PGXX22540</td>\n",
       "      <td>2</td>\n",
       "      <td>27.1</td>\n",
       "      <td>99.005</td>\n",
       "      <td>XK</td>\n",
       "      <td>33788</td>\n",
       "      <td>33986</td>\n",
       "      <td>66780</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>PGXX22540#2#27.1#XK</td>\n",
       "      <td>PGXX22540#2#27.1#XK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44349</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>PLPN239131</td>\n",
       "      <td>1</td>\n",
       "      <td>38.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>XK</td>\n",
       "      <td>16390</td>\n",
       "      <td>16590</td>\n",
       "      <td>44399</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>PLPN239131#1#38.1#XK</td>\n",
       "      <td>PLPN239131#1#38.1#XK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44350</th>\n",
       "      <td>XK</td>\n",
       "      <td>5</td>\n",
       "      <td>PLPN239131</td>\n",
       "      <td>2</td>\n",
       "      <td>38.1</td>\n",
       "      <td>99.010</td>\n",
       "      <td>XK</td>\n",
       "      <td>16381</td>\n",
       "      <td>16581</td>\n",
       "      <td>44413</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>PLPN239131#2#38.1#XK</td>\n",
       "      <td>PLPN239131#2#38.1#XK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>44351 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      anchorgene pos      sample hap contig  identity contiggene  tstart  \\\n",
       "0         A4GALT   3     GM19467   2   60.0   100.000     A4GALT  191956   \n",
       "1         A4GALT   3     GM19471   2   21.0   100.000     A4GALT  165882   \n",
       "2         A4GALT   3     GM19473   1    6.0   100.000     A4GALT  118198   \n",
       "3         A4GALT   3     GM19473   2    6.0   100.000     A4GALT  117969   \n",
       "4         A4GALT   3     HG00097   1   47.0   100.000     A4GALT  107008   \n",
       "...          ...  ..         ...  ..    ...       ...        ...     ...   \n",
       "44346         XK   5   PGXX22539   1    9.0    99.005         XK   30322   \n",
       "44347         XK   5   PGXX22539   2    9.0    99.005         XK   30314   \n",
       "44348         XK   5   PGXX22540   2   27.1    99.005         XK   33788   \n",
       "44349         XK   5  PLPN239131   1   38.1   100.000         XK   16390   \n",
       "44350         XK   5  PLPN239131   2   38.1    99.010         XK   16381   \n",
       "\n",
       "         tend  length  gene_match  strand_match              newheader  \\\n",
       "0      192156  288477        True          True  GM19467#2#60.0#A4GALT   \n",
       "1      166082  274006        True          True  GM19471#2#21.0#A4GALT   \n",
       "2      118398  280985        True          True   GM19473#1#6.0#A4GALT   \n",
       "3      118169  280754        True          True   GM19473#2#6.0#A4GALT   \n",
       "4      107208  212619        True          True  HG00097#1#47.0#A4GALT   \n",
       "...       ...     ...         ...           ...                    ...   \n",
       "44346   30521  105697        True          True     PGXX22539#1#9.0#XK   \n",
       "44347   30512  105714        True          True     PGXX22539#2#9.0#XK   \n",
       "44348   33986   66780        True          True    PGXX22540#2#27.1#XK   \n",
       "44349   16590   44399        True          True   PLPN239131#1#38.1#XK   \n",
       "44350   16581   44413        True          True   PLPN239131#2#38.1#XK   \n",
       "\n",
       "                   oldheader  \n",
       "0      GM19467#2#60.0#A4GALT  \n",
       "1      GM19471#2#21.0#A4GALT  \n",
       "2       GM19473#1#6.0#A4GALT  \n",
       "3       GM19473#2#6.0#A4GALT  \n",
       "4      HG00097#1#47.0#A4GALT  \n",
       "...                      ...  \n",
       "44346     PGXX22539#1#9.0#XK  \n",
       "44347     PGXX22539#2#9.0#XK  \n",
       "44348    PGXX22540#2#27.1#XK  \n",
       "44349   PLPN239131#1#38.1#XK  \n",
       "44350   PLPN239131#2#38.1#XK  \n",
       "\n",
       "[44351 rows x 14 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "raw",
   "id": "d91f1cc7-86b9-4e78-9d5e-a660f7560994",
   "metadata": {},
   "source": [
    "Export dataframe to correct headers with genematch and orientation with strandmatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f0587d36-af1c-40c2-ba77-d10d8f9348a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_fasta = '/g/data/te53/sj2852/blgrp/analyses/pangenome/analysis-v2/labelled2chm13-blast2anchors/bldgrp.genes.combined.fasta'\n",
    "output_fasta = '/g/data/te53/sj2852/blgrp/analyses/pangenome/analysis-v2/labelled2chm13-blast2anchors/bldgrp.genes.combined.corrected.fasta'\n",
    "\n",
    "# Create a dictionary to map old headers to new headers and strand_match status manually\n",
    "header_mapping = {}\n",
    "for _, row in df.iterrows():\n",
    "    old_header = row['oldheader']\n",
    "    header_mapping[old_header] = {'newheader': row['newheader'], 'strand_match': row['strand_match']}\n",
    "\n",
    "    \n",
    "# Create a list to hold modified SeqRecords\n",
    "modified_records = []\n",
    "\n",
    "# Iterate over the input FASTA file\n",
    "with open(input_fasta, 'r') as fasta_in:\n",
    "    for record in SeqIO.parse(fasta_in, 'fasta'):\n",
    "        old_header = record.id\n",
    "        \n",
    "        if old_header in header_mapping:\n",
    "            new_header = header_mapping[old_header]['newheader']\n",
    "            strand_match = header_mapping[old_header]['strand_match']\n",
    "            \n",
    "            # Reverse complement the sequence if strand_match is False\n",
    "            if not strand_match:\n",
    "                modified_seq = record.seq.reverse_complement()\n",
    "            else:\n",
    "                modified_seq = record.seq\n",
    "            \n",
    "            # Create a new SeqRecord with the modified header and sequence\n",
    "            modified_record = SeqIO.SeqRecord(\n",
    "                modified_seq,\n",
    "                id=new_header,\n",
    "                description=''\n",
    "            )\n",
    "            modified_records.append(modified_record)\n",
    "\n",
    "# Write the modified sequences to a new FASTA file\n",
    "with open(output_fasta, 'w') as fasta_out:\n",
    "    SeqIO.write(modified_records, fasta_out, 'fasta')"
   ]
  },
  {
   "cell_type": "raw",
   "id": "d167b475-c066-4163-84e5-17c1b50d3b9c",
   "metadata": {},
   "source": [
    "Handling gene bodies (contigs not contained in paf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "515f69dc-1af9-4e25-b014-ee6fb83487ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load PAF file\n",
    "input_file = \"/g/data/te53/sj2852/blgrp/analyses/pangenome/analysis-v2/labelled2chm13-blast2anchors/bldgrp.flank.nodup.paf\"\n",
    "column_names_paf = ['geneanchor', 'contig', 'identity', 'length', 'mismatches', 'gaps', 'qstart', 'qend', 'tstart', 'tend', 'evalue', 'bitscore']\n",
    "paf = pd.read_csv(input_file, sep='\\t', header=None, names=column_names_paf)\n",
    "paf = paf[['geneanchor', 'contig', 'identity', 'tstart', 'tend']]\n",
    "\n",
    "# Load FAI file\n",
    "fai_file = \"/g/data/te53/sj2852/blgrp/analyses/pangenome/analysis-v2/labelled2chm13-blast2anchors/bldgrp.genes.combined.fasta.fai\"\n",
    "column_names_fai = ['contig', 'length', 'x', 'y', 'z']\n",
    "faidf = pd.read_csv(fai_file, sep='\\t', header=None, names=column_names_fai)\n",
    "faidf = faidf[['contig', 'length']]\n",
    "\n",
    "# Find contigs in FAI but not in PAF and exclude those containing 'geneUn'\n",
    "contigs_in_fai_not_paf = faidf[~faidf['contig'].isin(paf['contig'])]\n",
    "contigs_in_fai_not_paf = contigs_in_fai_not_paf[~contigs_in_fai_not_paf['contig'].str.contains(\"geneUn\")]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d4dc6198-d5bc-4a7a-a8af-69aa5e0bcd1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input FASTA file with all sequences\n",
    "fasta_input = \"/g/data/te53/sj2852/blgrp/analyses/pangenome/analysis-v2/labelled2chm13-blast2anchors/bldgrp.genes.combined.fasta\"\n",
    "\n",
    "# Output FASTA file to append the sequences to\n",
    "fasta_output = '/g/data/te53/sj2852/blgrp/analyses/pangenome/analysis-v2/labelled2chm13-blast2anchors/bldgrp.genes.combined.corrected.fasta'\n",
    "\n",
    "# Create a set of contigs to extract from the input FASTA\n",
    "contigs_to_extract = set(contigs_in_fai_not_paf['contig'])\n",
    "\n",
    "# Open the output FASTA file\n",
    "with open(fasta_output, \"a\") as output_handle:\n",
    "    # Parse the input FASTA file\n",
    "    for record in SeqIO.parse(fasta_input, \"fasta\"):\n",
    "        if record.id in contigs_to_extract:\n",
    "            # Write the record to the output FASTA\n",
    "            SeqIO.write(record, output_handle, \"fasta\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
